### ISSUE TITLE
Evolutionary algorithms for automatic agent strategy refinement

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#47]
## Summary

Implement evolutionary algorithms to automatically discover and refine agent strategies by breeding successful agents from the crowdsourced "known good" collection, evolving better parameter configurations over generations.

## Context

After building up a collection of proven agent configurations through distributed discovery (Issue #46), we can apply evolutionary algorithms to:

1. **Use known-good agents as seed population** - Start with strategies that already work
2. **Cross-breed successful traits** - Combine parameters from high-performers
3. **Mutate for exploration** - Introduce controlled variation
4. **Select for fitness** - Keep winners, discard losers
5. **Iterate over generations** - Progressively improve strategies

This transforms the platform from passive collection to active evolution of agent intelligence, all running in users' browsers.

## Research Questions

- What emergent strategies arise from evolutionary pressure?
- Do evolved agents discover novel cooperation patterns?
- Can evolution find counter-strategies to dominant agents?
- How does fitness landscape vary across scenarios?
- What parameter combinations are evolutionarily stable?

### Evolutionary Algorithm Components

**Population Management**
- Initialize from known-good agents registry
- Maintain population size (e.g., 50-100 agents)
- Track generation number and lineage
- Store fitness history across generations

**Fitness Evaluation**
- Run tournaments to measure agent performance
- Multiple fitness metrics:
  - Win rate (competitive fitness)
  - Average score (absolute fitness)
  - Robustness (performance variance)
  - Scenario diversity (generalization)
- Configurable fitness function weights

**Selection Strategies**
- Tournament selection (pick best from random subset)
- Roulette wheel (probability proportional to fitness)
- Elitism (always keep top N agents)
- Diversity preservation (maintain parameter variety)

**Crossover Operations**
- Single-point crossover (split parameter vector)
- Uniform crossover (randomly pick from each parent)
- Arithmetic crossover (blend parameter values)
- Respect parameter constraints during crossover

**Mutation Operations**
- Gaussian mutation (add random noise to parameters)
- Uniform mutation (random value within range)
- Adaptive mutation rate (higher early, lower late)
- Per-parameter mutation probability

### User Interface

**Evolution Setup**
```
┌─────────────────────────────────────────┐
│  Start Evolution Experiment             │
├─────────────────────────────────────────┤
│  Initial Population:                    │
│    ⚫ Load from Known Good Agents (50)   │
│    ⚪ Random Initialization              │
│    ⚪ Custom Selection                   │
│                                         │
│  Evolution Settings:                    │
│    Generations: [20] ────○──────        │
│    Population Size: [50]                │
│    Mutation Rate: [0.1] (10%)           │
│    Crossover Rate: [0.7] (70%)          │
│    Elite Size: [5] (top agents)         │
│                                         │
│  Fitness Evaluation:                    │
│    Games per Agent: [20]                │
│    Scenario: [Greedy Neighbor ▼]        │
│    Fitness: [Win Rate ▼]                │
│                                         │
│  Estimated time: ~15 minutes            │
│                                         │
│  [Cancel]  [Start Evolution]            │
└─────────────────────────────────────────┘
```

**Evolution Progress**
```
┌─────────────────────────────────────────┐
│  Evolution Running... Gen 7/20          │
├─────────────────────────────────────────┤
│  ▓▓▓▓▓▓▓░░░░░░░░░░░░                   │
│                                         │
│  Current Generation:                    │
│    Best Fitness: 0.78 (+0.12 from gen 1)│
│    Avg Fitness: 0.64                    │
│    Diversity: 0.45 (parameter variance) │
│                                         │
│  📈 Fitness Trend:                      │
│  0.8┤     ╭─╮                           │
│  0.6┤   ╭─╯ ╰─╮                         │
│  0.4┤╭──╯      ╰─                       │
│     └─────────────                      │
│     1  5  10  15  20                    │
│                                         │
│  [Pause] [Cancel]                       │
└─────────────────────────────────────────┘
```

**Evolution Results**
```
┌─────────────────────────────────────────┐
│  Evolution Complete! 🧬                 │
├─────────────────────────────────────────┤
│  Best Agent: Generation 18              │
│    Fitness: 0.82 (win rate)             │
│    Improvement: +32% from initial       │
│                                         │
│  📊 Evolution Statistics:               │
│    Generations: 20                      │
│    Total Evaluations: 1000 games        │
│    Convergence: Generation 15           │
│    Final Diversity: 0.38                │
│                                         │
│  🏆 Top 3 Evolved Agents:               │
│    1. Evolved Champion (0.82)           │
│    2. Strategic Variant (0.79)          │
│    3. Balanced Hybrid (0.77)            │
│                                         │
│  [View Lineage] [Submit to Registry]    │
│  [Download Population] [Run Again]      │
└─────────────────────────────────────────┘
```

### Visualization & Analysis

**Fitness Evolution Chart**
- Line chart showing best/average/worst fitness over generations
- Confidence intervals if multiple runs
- Convergence detection

**Parameter Space Exploration**
- 2D projection of parameter space (PCA/t-SNE)
- Show population distribution across generations
- Color by fitness value
- Animate evolution through generations

**Lineage Tree**
- Show parent-child relationships
- Highlight successful lineages
- Display parameter changes along branches
- Interactive exploration

**Parameter Heatmap**
- Show which parameters changed most during evolution
- Correlation between parameters and fitness
- Identify critical parameters for success

### Genetic Algorithm Core

```typescript
interface EvolutionConfig {
  populationSize: number;
  generations: number;
  mutationRate: number;
  crossoverRate: number;
  eliteSize: number;
  fitnessFunction: FitnessFunction;
  selectionStrategy: 'tournament' | 'roulette' | 'rank';
}

interface Individual {
  genome: AgentParameters;  // Parameter vector
  fitness?: number;
  generation: number;
  parents?: [string, string];  // IDs of parent individuals
  id: string;
}

class GeneticAlgorithm {
  async evolve(config: EvolutionConfig): Promise<Population> {
    let population = this.initializePopulation(config);

    for (let gen = 0; gen < config.generations; gen++) {
      // Evaluate fitness
      await this.evaluateFitness(population, config);

      // Selection
      const parents = this.selectParents(population, config);

      // Crossover
      const offspring = this.crossover(parents, config);

      // Mutation
      this.mutate(offspring, config);

      // Elitism + offspring = new population
      population = this.formNextGeneration(
        population,
        offspring,
        config
      );

      // Report progress
      this.emitProgress(gen, population);

      // Early stopping if converged
      if (this.hasConverged(population)) break;
    }

    return population;
  }

  private async evaluateFitness(
    population: Individual[],
    config: EvolutionConfig
  ): Promise<void> {
    // Run tournaments in parallel (Web Workers)
    const results = await Promise.all(
      population.map(ind =>
        this.runTournament(ind.genome, config.fitnessFunction)
      )
    );

    population.forEach((ind, i) => {
      ind.fitness = results[i];
    });
  }

  private crossover(
    parents: Individual[],
    config: EvolutionConfig
  ): Individual[] {
    const offspring: Individual[] = [];

    for (let i = 0; i < parents.length - 1; i += 2) {
      if (Math.random() < config.crossoverRate) {
        const [child1, child2] = this.uniformCrossover(
          parents[i],
          parents[i + 1]
        );
        offspring.push(child1, child2);
      } else {
        offspring.push(parents[i], parents[i + 1]);
      }
    }

    return offspring;
  }

  private uniformCrossover(
    parent1: Individual,
    parent2: Individual
  ): [Individual, Individual] {
    const child1: AgentParameters = {};
    const child2: AgentParameters = {};

    for (const key of Object.keys(parent1.genome)) {
      if (Math.random() < 0.5) {
        child1[key] = parent1.genome[key];
        child2[key] = parent2.genome[key];
      } else {
        child1[key] = parent2.genome[key];
        child2[key] = parent1.genome[key];
      }
    }

    return [
      { genome: child1, generation: parent1.generation + 1, ... },
      { genome: child2, generation: parent1.generation + 1, ... }
    ];
  }

  private mutate(
    population: Individual[],
    config: EvolutionConfig
  ): void {
    for (const individual of population) {
      for (const key of Object.keys(individual.genome)) {
        if (Math.random() < config.mutationRate) {
          // Gaussian mutation
          const noise = this.gaussianRandom(0, 0.1);
          individual.genome[key] = this.clamp(
            individual.genome[key] + noise,
            0, 1  // Parameter bounds
          );
        }
      }
    }
  }
}
```

### Parallel Fitness Evaluation

```typescript
// Use Web Workers for parallel tournament execution
class ParallelEvaluator {
  private workers: Worker[] = [];

  constructor(numWorkers: number = navigator.hardwareConcurrency || 4) {
    for (let i = 0; i < numWorkers; i++) {
      this.workers.push(new Worker('tournament-worker.js'));
    }
  }

  async evaluatePopulation(
    population: Individual[],
    gamesPerIndividual: number
  ): Promise<number[]> {
    const tasks = population.map((ind, idx) => ({
      individual: ind,
      gamesPerIndividual,
      taskId: idx
    }));

    // Distribute tasks across workers
    const results = await this.distributeWork(tasks);
    return results;
  }
}
```

### Diversity Preservation

```typescript
function calculateDiversity(population: Individual[]): number {
  // Measure parameter space variance
  const parameterVectors = population.map(ind =>
    Object.values(ind.genome)
  );

  // Calculate average pairwise distance
  let totalDistance = 0;
  let count = 0;

  for (let i = 0; i < parameterVectors.length; i++) {
    for (let j = i + 1; j < parameterVectors.length; j++) {
      totalDistance += euclideanDistance(
        parameterVectors[i],
        parameterVectors[j]
      );
      count++;
    }
  }

  return totalDistance / count;
}

function maintainDiversity(
  population: Individual[],
  minDiversity: number
): void {
  if (calculateDiversity(population) < minDiversity) {
    // Inject random individuals to increase diversity
    const numToInject = Math.floor(population.length * 0.1);
    for (let i = 0; i < numToInject; i++) {
      population[i] = createRandomIndividual();
    }
  }
}
```

## Acceptance Criteria

- [ ] GA successfully evolves population over multiple generations
- [ ] Fitness improves over time (best and average)
- [ ] Crossover operators produce valid offspring
- [ ] Mutation respects parameter constraints
- [ ] Elitism preserves top performers
- [ ] Evolution progress visualized in real-time
- [ ] Final population can be exported/submitted
- [ ] Lineage tracking works correctly
- [ ] Parallel evaluation speeds up fitness computation
- [ ] Diversity preservation prevents premature convergence
- [ ] Results can be analyzed (charts, statistics)
- [ ] Documentation includes example experiments

### Baseline Experiments
1. **Scenario-specific evolution**: Evolve agents for each scenario separately
2. **Multi-scenario evolution**: Evolve generalists that work across scenarios
3. **Competitive evolution**: Evolve agents to beat specific opponents

### Advanced Experiments
4. **Co-evolution**: Evolve teams of agents simultaneously
5. **Novelty search**: Reward behavioral diversity, not just fitness
6. **Multi-objective evolution**: Optimize for multiple goals (score, efficiency, fairness)
7. **Island model**: Multiple populations evolve separately, occasionally migrate

### Research Questions
- Do evolved agents rediscover known optimal strategies?
- Do they discover novel strategies humans missed?
- How does evolution navigate the exploration/exploitation tradeoff?
- What parameter correlations emerge naturally?

## Testing Strategy

**Unit Tests**:
- [ ] GA operators (selection, crossover, mutation)
- [ ] Parameter constraint validation during crossover/mutation
- [ ] Fitness calculation accuracy
- [ ] Diversity metrics computation
- [ ] Lineage tracking and parent IDs

**Integration Tests**:
- [ ] Full evolution run (small scale: 10 generations, 20 population)
- [ ] Parallel fitness evaluation with Web Workers
- [ ] Progress tracking and event emissions
- [ ] Export/import evolved populations
- [ ] Integration with tournament engine

**Performance Tests**:
- [ ] WASM integration (verify 10-20x speedup)
- [ ] Web Worker parallelization effectiveness
- [ ] Memory usage during long runs (no leaks)
- [ ] Time estimation accuracy
- [ ] Convergence detection

**End-to-End Tests**:
- [ ] Complete evolution workflow (setup → run → results)
- [ ] Visualization updates during evolution (real-time charts)
- [ ] Pause/resume functionality
- [ ] Submit evolved agents to registry (#46)
- [ ] Load evolved agents into team builder (#44)

## Dependencies

- [ ] #45 (Tournament runner) - Required for fitness evaluation (WASM performance critical)
- [ ] #46 (Agent collection) - Required for seed population of known-good agents
- [x] #44 (Team builder) - Required to import evolved agents

**Dependency Status**: Issues #45 and #46 must complete before this can proceed.

- #44: ✅ CLOSED
- #45: 🔨 IN PROGRESS (`loom:building`)
- #46: 🔴 BLOCKED (waiting on #45)
## Future Enhancements

- **Multi-population evolution** (island model)
- **Adaptive mutation rates** (higher when stuck, lower when improving)
- **Novelty search** (reward behavioral diversity)
- **MAP-Elites** (maintain archive of diverse high-performers)
- **Neuroevolution** (evolve neural network policies)
- **Co-evolution** (evolve competing populations simultaneously)
- **Evolution dashboard** (compare multiple evolution runs)
- **Community evolution challenges** (evolve agents to solve specific problems)

## Success Metrics

- Fitness improvement over generations (absolute and relative)
- Convergence speed (generations to plateau)
- Final population diversity
- Discovery of novel strategies
- Reproducibility (consistent results across runs)
- Performance vs. hand-tuned baseline agents

## Notes

Evolutionary algorithms are particularly well-suited for this platform because:

1. **Fitness evaluation is expensive** - running tournaments takes time, so we need sample-efficient evolution
2. **Parameter space is continuous** - real-valued parameters work well with genetic operators
3. **No gradient information** - can't use gradient descent, evolution is gradient-free
4. **Multiple local optima** - evolution can escape local minima through diversity
5. **Interpretable results** - evolved parameters are human-readable

This positions Bucket Brigade as both:
- A **research platform** for studying multi-agent evolution
- A **practical tool** for discovering effective cooperation strategies

The combination of crowdsourced discovery (Issue #46) and evolutionary algorithms creates a powerful virtuous cycle:
- Users discover good agents → feed into evolution → produce better agents → expand search space → repeat

**Comment by rjwalters** [comment]
## Curator Review - Enhanced with Testing & Dependency Tracking ✅

### Enhancements Made

**1. Testing Strategy** - Added comprehensive test plan:
- Unit tests (GA operators, validation, metrics)
- Integration tests (full evolution run, parallelization)
- Performance tests (WASM speedup verification, memory usage)
- E2E tests (complete workflow, visualizations)

**2. Dependencies Section** - Updated to task list format:
- [x] #44 (Team builder) - COMPLETED ✅
- [ ] #45 (Tournament runner) - IN PROGRESS (labeled `loom:building`)
- [ ] #46 (Agent collection) - BLOCKED (waiting on #45)

**3. Dependency Status Tracking** - Added clear status indicators showing the dependency chain

### Curation Assessment

This issue demonstrates **exceptional research quality**:
- ✅ Clear scientific motivation with research questions
- ✅ Comprehensive GA components (selection, crossover, mutation)
- ✅ Detailed UI mockups for all evolution phases
- ✅ Extensive technical implementation guidance
- ✅ Multiple visualization approaches documented
- ✅ Experiment catalog for validation
- ✅ Thorough testing strategy (newly added)
- ✅ Performance analysis and WASM dependency clearly explained

### Blocking Status

**Current Status**: 🔴 **BLOCKED ON MULTIPLE DEPENDENCIES**

This issue has a dependency chain:
1. #45 (Tournament runner with WASM) must complete first
2. #46 (Agent collection) depends on #45
3. #47 (Evolution) depends on both #45 AND #46

**Critical Performance Dependency**: The existing Curator comment correctly identifies that **WASM is absolutely critical** for this feature:
- Without WASM: ~30 minutes per evolution run (impractical)
- With WASM + parallelization: ~25-30 seconds (practical!)
- This represents a **40-160x speedup**

**Recommendation**: 
- Label this issue as `loom:blocked`
- Wait for #45 AND #46 to complete
- This is the "crown jewel" feature but requires the full stack

### Builder Guidance (when unblocked)

The existing Curator comment provides outstanding implementation guidance:
- **Recommendation**: Custom GA implementation (~400 lines, TypeScript native)
- **Architecture**: Clean separation of GA engine, operators, and evaluator
- **Performance**: Parallel evaluation with Web Workers + WASM
- **Estimated effort**: 6-8 days
- **Phasing**: Core GA → Visualization → Polish

### Research Platform Value

This feature positions Bucket Brigade as a serious research platform:
- Distributed parameter space exploration (#46)
- Automated strategy discovery (this issue)
- Reproducible experiments
- Community-driven evolution
- Publication-quality visualizations

**Scientific Impact**: This could generate real research papers on:
- Emergent cooperation strategies in multi-agent systems
- Evolutionary stable strategies in social dilemmas
- Parameter space topology of cooperation problems

This issue is **ready for implementation** once dependencies are satisfied.
